FindMKL.cmake

```diff
@@ -0,0 +1,227 @@
+^^^^^^^^^^
+
+``SCALAPACK``
+
+
+Imported targets 
+^^^^^^^^^^^^^^^^
+
+``MKL::MKL``
+
+``MKL::SCALAPACK``
+
+
+Result variables
+^^^^^^^^^^^^^^^^
+
+``MKL_FOUND``
+  Found MKL.
+
+``MKL_SCALAPACK_FOUND``
+  Found ScaLAPACK.
+
+Not supported
+^^^^^^^^^^^^^
+
+- TBB threading back-end
+- F95 interfaces
+
+Note: Mixing GCC and Intel OpenMP backends is a bad idea.
+
+#]=======================================================================]
+
+# Modules
+#
+include(FindPackageHandleStandardArgs)
+
+# Functions
+#
+function(__mkl_find_library _name)
+    find_library(${_name}
+        NAMES ${ARGN}
+        HINTS ${MKL_ROOT}
+        PATH_SUFFIXES ${_mkl_libpath_suffix}
+                      lib
+        )
+    mark_as_advanced(${_name})
+endfunction()
+
+# Options
+#
+# The `NOT DEFINED` guards on CACHED variables are needed to make sure that 
+# normal variables with of same name always take precedence*.
+#
+# * There are many caveats with CACHE variables in CMake. Before version 
+#   3.12, both the `option()` command and `set(... CACHE ...)` would 
+#   override normal variables if cached equivalents don't exist or they 
+#   exisit but their type is not specified (e.g. command line arguments: 
+#   -DFOO=ON instead of -DFOO:BOOL=ON). For 3.13 with policy CMP0077, 
+#   `option()` no longer overrides normal variables of the same name.
+#   `set(... CACHE ...)` is still stuck with the old behaviour. 
+#
+#   https://cmake.org/cmake/help/v3.15/command/set.html#set-cache-entry
+#   https://cmake.org/cmake/help/v3.15/command/set.html#set-cache-entry
+#
+if(NOT DEFINED MKL_ROOT)
+    set(MKL_ROOT $ENV{MKLROOT} CACHE PATH "MKL's root directory.")
+endif()
+
+if(NOT DEFINED MKL_PARALLEL)
+    set(MKL_PARALLEL ON CACHE BOOL "Toggles parallel and sequential versions.")
+endif()
+
+if(NOT DEFINED MKL_64BIT)
+    set(MKL_64BIT OFF CACHE BOOL "Toggles 32bit/64bit MKL interfaces.")
+endif()
+
+# Determine MKL's library folder
+#
+set(_mkl_libpath_suffix "lib/intel64")
+if(CMAKE_SIZEOF_VOID_P EQUAL 4) # 32 bit
+    set(_mkl_libpath_suffix "lib/ia32")
+endif()
+
+if (WIN32)
+    string(APPEND _mkl_libpath_suffix "_win")
+elseif (APPLE)
+    string(APPEND _mkl_libpath_suffix "_mac")
+else ()
+    string(APPEND _mkl_libpath_suffix "_lin")
+endif ()
+
+# Determine 32bit or 64bit interface is used
+#
+set(_mkl_lp "lp64")
+if(MKL_64BIT)
+    set(_mkl_lp "ilp64")
+endif()
+
+# Find MKL header
+#
+find_path(MKL_INCLUDE_DIR mkl.h
+    HINTS ${MKL_ROOT}/include
+    )
+mark_as_advanced(MKL_INCLUDE_DIR)
+
+# Find MKL core libraries
+#
+__mkl_find_library(MKL_CORE_LIB mkl_core)
+__mkl_find_library(MKL_INTERFACE_LIB mkl_intel_${_mkl_lp})
+
+if (MKL_PARALLEL)
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NOT APPLE)
+        __mkl_find_library(MKL_THREADING_LIB mkl_gnu_thread)
+    else ()
+        __mkl_find_library(MKL_THREADING_LIB mkl_intel_thread)
+    endif ()
+else ()
+    __mkl_find_library(MKL_THREADING_LIB mkl_sequential)
+endif ()
+
+# Find ScaLAPACK
```

Comment:
I propose to open a PR on this branch with as objective the management of components.

I've just opened #3 for keeping track of this.

Reply:
Is there something wrong with how I manager components here?

Currently the module searches for core MKL, i.e. BLAS/LAPACK. The ScaLAPACK component is searched for and made available if found even if the component is not explicitly specified. I feel like having BLAS/LAPACK as a component is strange as it is the core of MKL and without it, none of the other components (i.e. ScaLAPACK or FFT (not yet implemented)) will work anyway.